src/Background/NotificationRequest.vala

```diff
@@ -24,95 +52,86 @@ public class NotificationRequest : Object {
         FAILED
     }
 
-    [DBus (name = "org.freedesktop.Notifications")]
-    public interface Notifications : Object {
-        public signal void action_invoked (uint32 id, string action_key);
-        public signal void notification_closed (uint32 id, uint32 reason);
-
-        public abstract void close_notification (uint32 id) throws DBusError, IOError;
-        public abstract uint32 notify (
-            string app_name,
-            uint32 replaces_id,
-            string app_icon,
-            string summary,
-            string body,
-            string[] actions,
-            HashTable<string, Variant> hints,
-            int32 expire_timeout
-        ) throws DBusError, IOError;
+    static construct {
+        requests = new HashTable<uint32, unowned NotificationRequest> (null, null);
     }
 
-    [DBus (visible = false)]
-    public static void init (DBusConnection connection) {
-        notification_by_id = new HashTable<uint32, NotificationRequest> (null, null);
-
-        try {
-            notifications = connection.get_proxy_sync ("org.freedesktop.Notifications", "/org/freedesktop/Notifications");
-            notifications.action_invoked.connect (on_action_invoked);
-            notifications.notification_closed.connect (on_notification_closed);
-        } catch {
-            warning ("Cannot connect to notifications dbus, background portal working with reduced functionality.");
-        }
-    }
-
-    private static void on_action_invoked (uint32 id, string action_key) {
-        var notification = notification_by_id.take (id);
+    private static void action_invoked (uint32 id, string action_key) {
+        unowned var notification = requests.take (id);
         if (notification == null) {
             return;
         }
 
         if (action_key == ACTION_ALLOW_BACKGROUND) {
-            notification.response (NotifyBackgroundResult.ALLOW);
+            notification.response (ALLOW);
         } else {
-            notification.response (NotifyBackgroundResult.FORBID);
+            notification.response (FORBID);
         }
     }
 
-    private static void on_notification_closed (uint32 id, uint32 reason) {
-        var notification = notification_by_id.take (id);
+    private static void notification_closed (uint32 id, Fdo.Notifications.CloseReason reason) {
+        unowned var notification = requests.take (id);
         if (notification == null) {
             return;
         }
 
-        if (reason == 2) { //Dismissed by user
-            notification.response (NotifyBackgroundResult.ALLOW_ONCE);
-        } else if (reason == 3) { //Closed via DBus call
-            notification.response (NotifyBackgroundResult.CANCELLED);
+        if (reason == CANCELLED) { // Closed via DBus call
+            notification.response (CANCELLED);
+        } else { // Dismissed, Expired, or something internal to the server
```

Comment:
A expired signal is also sent when the notification bubble closes. I think it's better to keep waiting then because the notification will stay in the indicator and can still send a response when something like elementary/wingpanel-indicator-notifications#258 gets merged.

Reply:
This's a notification server bug, currently the server and indicator works at the back of each other, making the integration between both a mess. I'm working in a "rewrite/refactor" of the server right now, so that we can support the indicator properly.

we can wait till the server is fixed, but i don't see why, because the expiration would make the action activation useless in the server side. and we would need to fix that before something like elementary/wingpanel-indicator-notifications#258 is merged.

Reply:
Only out of curiosity: Why would the expiration make the action activation useless even if we ignore the expiration signal?

Reply:
If the server emits a NotificationClosed signal it's mean that the notification doesn't exists anymore. so there will be no other signal emission for that notification id from the server.